dissect/squashfs/squashfs.py

```diff
@@ -217,19 +219,34 @@ def header(self) -> Instance:
         base_struct = c_squashfs.squashfs_base_inode_header
 
         block = self.fs.sb.inode_table_start + self.block
-        _, _, data = self.fs._read_metadata(block, self.offset, len(base_struct))
+        data_block, data_offset, data = self.fs._read_metadata(block, self.offset, len(base_struct))
 
         header = base_struct(data)
         actual_struct = INODE_STRUCT_MAP.get(header.inode_type)
 
         if len(actual_struct) != len(base_struct):
-            _, _, data = self.fs._read_metadata(block, self.offset, len(actual_struct))
+            data_block, data_offset, data = self.fs._read_metadata(block, self.offset, len(actual_struct))
 
         if actual_struct != base_struct:
             header = actual_struct(data)
 
+        self._data_block = data_block
+        self._data_offset = data_offset
+
         return header
 
+    @cached_property
+    def data_block(self) -> int:
+        if not self._data_block:
+            self.header  # This will set self._data_block
```

Comment:
Good comment, but would it not be better to separate this into a different function then? Now this property has a surprising side-effect.

Reply:
Given how `@cached_property` works, I guess the cleanest  way is to stack these functions like this:

```
def _metadata(self):
    ...

    self.data_block = data_block
    self.data_offset = data_offset
    self.header = header

    return data_block, data_offset, header

@cached_property
def data_block(self):
    data_block, _, _ = self._metadata
    return data_block

@cached_property
def data_offset(self):
    _, data_offset, _ = self._metadata
    return data_offset

@cached_property
def header(self):
    _, _, header = self._metadata
    return header
```

Functions decorated with `@cached_property` will be called only if the property name is not present in `self.__dict__`. When called, the property name and return value of the function are added to `self.__dict__`, masking the decorated function in `self.__class__._dict__`.

So in this example the logic in `def _metadata()` will be run only once for whichever of the `data_block()` `data_offset()` or `header()` functions is called first. After that, these functions are transformed to normal properties. As a bonus `self._data_block` and `self._data_offset` are not needed.

Reply:
Thanks @pyrco, nice suggestion!